src/ipc/dbus/DBusObjectRegistry.cpp

```diff
@@ -66,56 +77,125 @@ void DBusObjectRegistry::registerObject(const QString &objectPath, facelift::Asy
     }
 }
 
 void DBusObjectRegistry::unregisterObject(const QString &objectPath)
 {
     init();
     auto serviceName = DBusManager::instance().serviceName();
-    if (m_master) {
+    if (isMaster()) {
         m_master->unregisterObject(objectPath, serviceName);
     } else {
-        return m_objectRegistryAsyncProxy->unregisterObject(objectPath, serviceName);
+        m_objectRegistryAsyncProxy->unregisterObject(objectPath, serviceName);
+    }
+}
+
+void DBusObjectRegistry::updateObjects(const QMap<QString, QString>& objectMap)
+{
+    int version = objectMap[VERSION_KEY].toInt(); // 0 if key is not present
+    Q_ASSERT(version != INVALID_REGISTRY_VERSION);
+    if (!hasValidObjects()) {
+        m_registryVersion = version;
+        m_objects = objectMap;
+        m_objects.remove(VERSION_KEY);
+        emit objectsChanged();
+    }
+}
+
+int DBusObjectRegistry::nextVersion(const int currentVersion)
+{
+    int version = currentVersion + 1;
+    if (version == std::numeric_limits<int>::max()) { // overflow protection
+        version = INITIAL_REGISTRY_VERSION;
     }
+    return version;
 }
 
-const QMap<QString, QString> &DBusObjectRegistry::objects(bool blocking)
+void DBusObjectRegistry::syncObjects()
+{
+    ObjectRegistryIPCDBusProxy objectRegistryProxy;
+    objectRegistryProxy.ipc()->setServiceName(m_serviceName);
+    objectRegistryProxy.connectToServer();
+    updateObjects(objectRegistryProxy.getObjects());
+}
+
+const QMap<QString, QString>& DBusObjectRegistry::objects(bool blocking)
 {
     init();
-    if (m_master) {
-        return m_master->objects();
+    if (!isMaster() && blocking && !hasValidObjects()) {
+        syncObjects();
+    }
+    return m_objects;
+}
+
+void DBusObjectRegistry::onObjectAdded(const QString &objectPath, const QString &serviceName, int registryVersion)
+{
+    if (hasValidObjects() && registryVersion == nextVersion(m_registryVersion)) {
+        m_registryVersion = registryVersion;
+        if (!m_objects.contains(objectPath)) {
+            m_objects.insert(objectPath, serviceName);
+            emit objectsChanged();
+        } else {
+            qCCritical(LogIpc) << "Cannot add object with object path:" << objectPath << ", service name:" << serviceName
+                               << "Object registry already contains this object path with service name:" << m_objects[objectPath];
+        }
     } else {
-        if (blocking) {
-            if (m_objectRegistryProxy == nullptr) {
-                m_objectRegistryProxy = new facelift::ipc::dbus::ObjectRegistryIPCDBusProxy();
-                m_objectRegistryProxy->ipc()->setServiceName(m_serviceName);
-                m_objectRegistryProxy->connectToServer();
-            }
-            return m_objectRegistryProxy->objects();
+        qCWarning(LogIpc) << "Unexpected object registry version. Object path:" << objectPath << ", service name:" << serviceName;
```

Comment:
this is also executed when hasValidObjects() is false.

Reply:
That is expected. There is no correct version in this situation.
Please let me know if you expect any specific behavior or just different message.